### ISSUE TITLE
Istio 1.6.6 breaks auto mTLS for Pod to Pod (by PodIP) for Pod with ServiceEntry MESH_INTERNAL

### ISSUE THREAD

**Issue by henrysfx** [istio/istio#27042]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**

Istio 1.6.6 breaks networking with auto mTLS enabled between pods within mesh (with sidecar) when making requests via pod IP to pod which has a `location:MESH_INTERNAL` and `resolution:NONE` service entry.   This seems like the likely culprit: https://github.com/istio/istio/commit/6779bb6247ab64475eba2e099bfa33fb94cad11a
Also posted in: https://discuss.istio.io/t/istio-1-6-6-breaks-auto-mtls-for-pod-to-pod-by-podip/7717

**Affected product area (please put an X in all that apply)**

[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**

To continue to be able to make requests to pods by IP that have a `location:MESH_INTERNAL` and `resolution:NONE` ServiceEntry with mTLS

**Steps to reproduce the bug**

* Install istio 1.6.6
* Create a server pod with exposed port (ensure istio sidecar is injected)
* Create a client pod with exposed port (ensure istio sidecar is injected)
* Create a serviceEntry like this:
```
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  annotations:
  name: myserverpod
spec:
  addresses:
  - <server pod ip>
  hosts:
  - <server pod name>
  location: MESH_INTERNAL
  ports:
  - name: myserverport
    number: <server port>
    protocol: TCP
  resolution: NONE
```
* test making a request to the server podIP:port from the client pod

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.6.6

**How was Istio installed?**
istioctl

**Environment where bug was observed (cloud vendor, OS, etc)**
GKE

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.
See comparison of cluster output from 1.6.5 (working) to 1.6.6 (broken)
1.6.5
```
  {
        "name": "outbound|8080||test-server-6f5bc8bd5c-l7xbt",
        "type": "ORIGINAL_DST",
        "connectTimeout": "10s",
        "lbPolicy": "CLUSTER_PROVIDED",
        "circuitBreakers": {
            "thresholds": [
                {
                    "maxConnections": 4294967295,
                    "maxPendingRequests": 4294967295,
                    "maxRequests": 4294967295,
                    "maxRetries": 4294967295
                }
            ]
        },
        "transportSocket": {
            "name": "envoy.transport_sockets.tls",
            "typedConfig": {
                "@type": "type.googleapis.com/envoy.api.v2.auth.UpstreamTlsContext",
                "commonTlsContext": {
                    "tlsCertificateSdsSecretConfigs": [
                        {
                            "name": "default",
                            "sdsConfig": {
                                "apiConfigSource": {
                                    "apiType": "GRPC",
                                    "grpcServices": [
                                        {
                                            "envoyGrpc": {
                                                "clusterName": "sds-grpc"
                                            }
                                        }
                                    ]
                                }
                            }
                        }
                    ],
                    "combinedValidationContext": {
                        "defaultValidationContext": {},
                        "validationContextSdsSecretConfig": {
                            "name": "ROOTCA",
                            "sdsConfig": {
                                "apiConfigSource": {
                                    "apiType": "GRPC",
                                    "grpcServices": [
                                        {
                                            "envoyGrpc": {
                                                "clusterName": "sds-grpc"
                                            }
                                        }
                                    ]
                                }
                            }
                        }
                    },
                    "alpnProtocols": [
                        "istio-peer-exchange",
                        "istio"
                    ]
                },
                "sni": "outbound_.8080_._.test-server-6f5bc8bd5c-l7xbt"
            }
        },
        "filters": [
            {
                "name": "istio.metadata_exchange",
                "typedConfig": {
                    "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                    "typeUrl": "type.googleapis.com/envoy.tcp.metadataexchange.config.MetadataExchange",
                    "value": {
                        "protocol": "istio-peer-exchange"
                    }
                }
            }
        ]
    },
```

1.6.6
```
    {
        "name": "outbound|8080||test-server-6f5bc8bd5c-w847r",
        "type": "ORIGINAL_DST",
        "connectTimeout": "10s",
        "lbPolicy": "CLUSTER_PROVIDED",
        "circuitBreakers": {
            "thresholds": [
                {
                    "maxConnections": 4294967295,
                    "maxPendingRequests": 4294967295,
                    "maxRequests": 4294967295,
                    "maxRetries": 4294967295
                }
            ]
        },
        "filters": [
            {
                "name": "istio.metadata_exchange",
                "typedConfig": {
                    "@type": "type.googleapis.com/udpa.type.v1.TypedStruct",
                    "typeUrl": "type.googleapis.com/envoy.tcp.metadataexchange.config.MetadataExchange",
                    "value": {
                        "protocol": "istio-peer-exchange"
                    }
```

**Comment by howardjohn** [comment]
I think if you define the ServiceAccount field on the ServiceEntry it will work - otherwise we don't know what to use for mtls.

Maybe we could derive it from the address but not sure we can reliably do that

cc @hzxuzhonghu @incfly

**Comment by incfly** [comment]
> I think if you define the ServiceAccount field on the ServiceEntry it will work - otherwise we don't know what to use for mtls.

Are you suggesting we use the service account field as an indication that we should enable client side mTLS? (which is needed for abitrary manual service entry anyway I agree).

**Comment by howardjohn** [comment]
We are doing this https://github.com/istio/istio/blob/22d9a05c0f85a9da74e17211d50087a7dc713dc5/pilot/pkg/serviceregistry/serviceentry/conversion.go#L329

**Comment by incfly** [comment]
@hzxuzhonghu Also i remembered your change is only for headeless service. but in the PR, we use the `ORIGINAL_DST` typed cluster to infer that's headless service.

1. When headless pod is used, we have one outbound cluster for all pods, with ORIGIN_DST cluster. We disable the mTLS client config since the `matches` config can't work in a none EDS typed cluste.r
1. In this customer case, the service entry defines the discovery, we have one outbound cluster per pod IP.

It's surprising and interesting to me we have some code path to trigger one pod IP per cluster in case 2. If we are able to do it for headless as well, then we can just check whether pod has `security.istio.io/tlsMode: istio` label to configure `transport_socket`.

As for this issue, @henrysfx curious what's the reason to use service entry with manual pod IP config within cluster? Because you need to build some service discovery automation yourself?

**Comment by hzxuzhonghu** [comment]
Same curiosity from me. The pod ip can change when rolling update/restart

**Comment by hzxuzhonghu** [comment]
> If we are able to do it for headless as well, then we can just check whether pod has security.istio.io/tlsMode: istio label to configure transport_socket.

This is feasibe

**Comment by hzxuzhonghu** [comment]
A workaround is: declare mtls with a destinationrule

**Comment by henrysfx** [comment]
> As for this issue, @henrysfx curious what's the reason to use service entry with manual pod IP config within cluster? Because you need to build some service discovery automation yourself?

@incfly @hzxuzhonghu Yes, your assumption is correct.  Unfortunately, we don't currently use k8s services and do discovery ourselves and we wanted to take advantage of mTLS now until we can eventually move to k8s services with native mTLS service support in istio.  So basically, we are watching pod events and creating a service entry per pod to allow this to workaround that fact.  How terrible of an idea that is, we are still in the process of testing.  Would love to hear thoughts and concerns with the scalability of this as well.

> I think if you define the ServiceAccount field on the ServiceEntry it will work - otherwise we don't know what to use for mtls.
> A workaround is: declare mtls with a destinationrule

@howardjohn @hzxuzhonghu I thought about this but I believe both WorkloadEntry and DestinationRule address/host fields only allows for DNS valid names and we are referencing pod IPs

**Comment by howardjohn** [comment]
Service Entry has a service account field. Workload entry is a subset of service entry

**Comment by howardjohn** [comment]
also you should be able to do the dest rule, out the hostname not the ip like test-server-6f5bc8bd5c-l7xbt

**Comment by hzxuzhonghu** [comment]
I think we can close this, It is solved by #27105
